src/functions/env.ts

```diff
@@ -264,44 +264,72 @@ export function writeUserEnvs(toWrite: Record<string, string>, envOpts: UserEnvs
-  const envFiles = findEnvfiles(functionsSource, projectId, projectAlias, isEmulator);
-  const projectScopedFileName = `.env.${projectId}`;
 
-  const projectScopedFileExists = envFiles.includes(projectScopedFileName);
-  if (!projectScopedFileExists) {
-    createEnvFile(envOpts);
+  // Determine which .env file to write to, and create it if it doesn't exist
+  const allEnvFiles = findEnvfiles(functionsSource, projectId, projectAlias, isEmulator);
+  const targetEnvFile = envOpts.isEmulator
+    ? FUNCTIONS_EMULATOR_DOTENV
+    : `.env.${envOpts.projectId}`;
+  const targetEnvFileExists = allEnvFiles.includes(targetEnvFile);
+  if (!targetEnvFileExists) {
+    fs.writeFileSync(path.join(envOpts.functionsSource, targetEnvFile), "", { flag: "wx" });
+    logBullet(
+      clc.yellow(clc.bold("functions: ")) +
+        `Created new local file ${targetEnvFile} to store param values. We suggest explicitly adding or excluding this file from version control.`
+    );
   }
 
-  const currentEnvs = loadUserEnvs(envOpts);
+  // Throw if any of the keys are duplicate (note special case if emulator) or malformed
+  const fullEnvs = loadUserEnvs(envOpts);
+  const prodEnvs = isEmulator
+    ? loadUserEnvs({ ...envOpts, isEmulator: false })
+    : loadUserEnvs(envOpts);
+  checkForDuplicateKeys(isEmulator || false, Object.keys(toWrite), fullEnvs, prodEnvs);
   for (const k of Object.keys(toWrite)) {
     validateKey(k);
-    if (currentEnvs.hasOwnProperty(k)) {
-      throw new FirebaseError(
-        `Attempted to write param-defined key ${k} to .env files, but it was already defined.`
-      );
-    }
   }
 
+  // Write all the keys in a single filesystem access
   logBullet(
-    clc.cyan(clc.bold("functions: ")) +
-      `Writing new parameter values to disk: ${projectScopedFileName}`
+    clc.cyan(clc.bold("functions: ")) + `Writing new parameter values to disk: ${targetEnvFile}`
   );
+  let lines = "";
   for (const k of Object.keys(toWrite)) {
-    fs.appendFileSync(
-      path.join(functionsSource, projectScopedFileName),
-      formatUserEnvForWrite(k, toWrite[k])
-    );
+    lines += formatUserEnvForWrite(k, toWrite[k]);
   }
+  fs.appendFileSync(path.join(functionsSource, targetEnvFile), lines);
 }
 
-function createEnvFile(envOpts: UserEnvsOpts): string {
-  const fileToWrite = envOpts.isEmulator ? FUNCTIONS_EMULATOR_DOTENV : `.env.${envOpts.projectId}`;
-  logger.debug(`Creating ${fileToWrite}...`);
-
-  fs.writeFileSync(path.join(envOpts.functionsSource, fileToWrite), "", { flag: "wx" });
-  return fileToWrite;
+/**
+ * Errors if any of the provided keys are aleady defined in the .env fields.
+ * This seems like a simple presence check, but...
+ *
+ * For emulator deploys, it's legal to write a key to .env.local even if it's
+ * already defined in .env.projectId. This is a special case designed to follow
+ * the principle of least surprise for emulator users.
+ */
+export function checkForDuplicateKeys(
```

Comment:
Forgive my ignorance, but aren't we only prompting and writing params that weren't defined? Should the errors in this method be user-facing errors, or assertion failures that exit with code 2 (internal error)